### DISCUSSION TITLE
Links from a CMS or external api

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#24821] [Help]
I am using headless WordPress but this could probably apply to other API's too. When you create a link in the content editor and display it in Next.js it is a standard href like

`<a href="/some-link">click me</a>`

This is obviously a problem because when a user clicks on it, it is going to that page and reloading as opposed to using 

```
        <Link href="/some-link">
          <a>Click me</a>
        </Link>
```

To output all of my html from the API I use

```
                    <div
                      className="price-desc"
                      dangerouslySetInnerHTML={{
                        __html: data.content,
                      }}
                    />
```

Is it possible to somehow get all normal links to be converted to a Next.js link or what is the solution here because it is a problem for internal linking.

**Comment by maksimLobachevskiy** [comment]
Hi, I have the same problem, have you already found the solution? Will be grateful for sharing.

**Comment by ghost** [comment]
Hi there, sadly no solution. Vercel support said that it could be that my web host where the wordpress site is, was blocking IP addresses from where the image transformations happen (Imgix). My web host confirmed that they are not and this doesn’t happen with all my images, only some so I really don’t know. I have basically given up and am just using a bootstrap image instead with React Bootstrap. Maybe I will look at using Cloudinary or something similar in future, not sure yet.

**Reply by maksimLobachevskiy** [reply]
Hi, thanks for the reply, I meant about normal links to be converted to a Next.js link ))

**Comment by ghost** [comment]
Oh, sorry! Haha, I thought this was my image issue. Sorry, no solution on
this either. I’m just not winning in general I’m afraid!

Kind regards
Jonathan



----------
On Wed, Sep 8, 2021 at 2:53 PM, Lobachevskyi Maksym <
***@***.***> wrote:

Hi, thanks for the reply, I meant about normal links to be converted to a
Next.js link ))

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/vercel/next.js/discussions/24821#discussioncomment-1296115>,
or unsubscribe
<https://github.com/notifications/unsubscribe-auth/AG7HWLNFE5YGYKLE3F2DPJ3UA5MGNANCNFSM44EY23FA>
.
Triage notifications on the go with GitHub Mobile for iOS
<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
or Android
<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

**Comment by jnahumphreys** [comment]
I've been subscribed to this thread as I too was having this issue...

**TLDR:**
I found Next introduced a lot of work that Gatsby provides out of the box; namely `gatsby-plugin-catch-links` and their specific [CMS plugins](https://www.gatsbyjs.com/docs/how-to/sourcing-data/headless-cms/). Obviously if you're forced to use Next.JS there is a solution, however IMO Gatsby provides a much better experience without this extra boilerplate.

**Solution:**
To resolve links and images within a body of HTML coming from an external source (typically a CMS like WordPress's content editor), you'll need to swap out the nodes `image` and `a` (albeit internal links) for the next react `Image` and `Link` components. There are two ways to do this:

1. Use [html-react-parser](https://www.npmjs.com/package/html-react-parser) to do the swap from the HTML fetched within `getStaticProps()`
2. Convert the HTML to an AST (abstract syntax tree) in node (i.e within `getStaticProps()`), pass this AST down to the page component as props and then use a package like [react-ast](https://www.npmjs.com/package/react-ast) to do the swap

Alternatively, if your CMS can provide the content as markdown, you can use [react-markdown](https://www.npmjs.com/package/react-markdown) to swap vanilla HTML tags for custom React components.

Hope that helps. :)
